File: spring-spqr/src/main/java/com/example/demo/gql/dataloaders/CommentsDataLoader.java
package com.example.demo.gql.dataloaders;

import com.example.demo.gql.types.Comment;
import com.example.demo.service.PostService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.dataloader.MappedBatchLoader;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

@RequiredArgsConstructor
@Component
@Slf4j
public class CommentsDataLoader implements MappedBatchLoader<String, List<Comment>> {
    final PostService postService;

    @Override
    public CompletionStage<Map<String, List<Comment>>> load(Set<String> keys) {
        List<Comment> comments = this.postService.getCommentsByPostIdIn(keys);

        Map<String, List<Comment>> mappedComments = new HashMap<>();
        keys.forEach(
                k -> mappedComments.put(k, comments
                        .stream()
                        .filter(c -> c.getPostId().equals(k)).toList())
        );
        log.info("mapped comments: {}", mappedComments);
        return CompletableFuture.supplyAsync(() -> mappedComments);
    }
}

File: dgs-kotlin-co/src/main/kotlin/com/example/demo/gql/dataloader/CommentsDataLoader.kt
package com.example.demo.gql.dataloader

import com.example.demo.gql.types.Comment
import com.example.demo.service.PostService
import com.netflix.graphql.dgs.DgsDataLoader
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.asCoroutineDispatcher
import kotlinx.coroutines.flow.toList
import kotlinx.coroutines.future.future
import org.dataloader.MappedBatchLoader
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.UUID
import java.util.concurrent.CompletionStage
import java.util.concurrent.Executors

@DgsDataLoader(name = "commentsLoader")
class CommentsDataLoader(val postService: PostService) : MappedBatchLoader<UUID, List<Comment>> {
    val loaderScope = CoroutineScope(Executors.newCachedThreadPool().asCoroutineDispatcher())
    companion object {
        val log: Logger = LoggerFactory.getLogger(CommentsDataLoader::class.java)
    }

    override fun load(keys: Set<UUID>): CompletionStage<Map<UUID, List<Comment>>> = loaderScope.future {
        val comments = postService.getCommentsByPostIdIn(keys.toList()).toList()
        val mappedComments: MutableMap<UUID, List<Comment>> = mutableMapOf()
        keys.forEach { mappedComments[it] = comments.filter { c -> c.postId == it } }
        log.info("mapped comments: {}", mappedComments)
        mappedComments
    }
}